pkg/runtime/wfengine/wfengine_test.go

```diff
@@ -368,22 +397,96 @@ func TestConcurrentActivityExecution(t *testing.T) {
 	})
 
 	ctx := context.Background()
-	client := startEngine(ctx, r)
+	client, engine := startEngine(ctx, r)
+	for _, opt := range GetTestOptions() {
+		t.Run(opt(engine), func(t *testing.T) {
+			id, err := client.ScheduleNewOrchestration(ctx, "ActivityFanOut")
+			if assert.NoError(t, err) {
+				metadata, err := client.WaitForOrchestrationCompletion(ctx, id)
+				if assert.NoError(t, err) {
+					assert.True(t, metadata.IsComplete())
+					assert.Equal(t, `["9","8","7","6","5","4","3","2","1","0"]`, metadata.SerializedOutput)
+
+					// Because all the activities run in parallel, they should complete very quickly
+					assert.Less(t, metadata.LastUpdatedAt.Sub(metadata.CreatedAt), 3*time.Second)
+				}
+			}
+		})
+	}
+}
+
+// TestContinueAsNewWorkflow verifies that a workflow can "continue-as-new" to restart itself with a new input.
+func TestContinueAsNewWorkflow(t *testing.T) {
+	r := task.NewTaskRegistry()
+	r.AddOrchestratorN("ContinueAsNewTest", func(ctx *task.OrchestrationContext) (any, error) {
+		var input int32
+		if err := ctx.GetInput(&input); err != nil {
+			return nil, err
+		}
 
-	id, err := client.ScheduleNewOrchestration(ctx, "ActivityFanOut")
-	if assert.NoError(t, err) {
-		metadata, err := client.WaitForOrchestrationCompletion(ctx, id)
-		if assert.NoError(t, err) {
-			assert.True(t, metadata.IsComplete())
-			assert.Equal(t, `["9","8","7","6","5","4","3","2","1","0"]`, metadata.SerializedOutput)
+		if input < 10 {
+			if err := ctx.CreateTimer(0).Await(nil); err != nil {
+				return nil, err
+			}
+			ctx.ContinueAsNew(input + 1)
+		}
+		return input, nil
+	})
+
+	ctx := context.Background()
+	client, engine := startEngine(ctx, r)
+	for _, opt := range GetTestOptions() {
+		t.Run(opt(engine), func(t *testing.T) {
+			id, err := client.ScheduleNewOrchestration(ctx, "ContinueAsNewTest", api.WithInput(0))
+			if assert.NoError(t, err) {
+				metadata, err := client.WaitForOrchestrationCompletion(ctx, id)
+				if assert.NoError(t, err) {
+					assert.True(t, metadata.IsComplete())
+					assert.Equal(t, `10`, metadata.SerializedOutput)
```

Comment:
Do we want to check the history size? I think there are other tests that validate the history gets reset on continuation so I don't know if we need to do it here as well, only if you think it's a useful signal.

Reply:
I'm a bit torn on this because it's a good thing to do but it's an implementation detail we don't have full control over. The actual history size is controlled by the `durabletask-go` dependency. The Dapr code is mostly trying to ensure that we faithfully represent this history in actor state. It's not straightforward to know what exactly to assert in terms of history size.